compiler-plugin/src/main/kotlin/arrow/meta/quotes/Parameter.kt

```diff
@@ -0,0 +1,31 @@
+package arrow.meta.quotes
+
+import arrow.meta.Meta
+import arrow.meta.phases.ExtensionPhase
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.modalityModifierType
+import org.jetbrains.kotlin.psi.psiUtil.visibilityModifierType
+
+/**
+ * A [KtParameter] [Quote] with a custom template destructuring [ParameterScope]
+ */
+fun Meta.parameter(
+  match: KtParameter.() -> Boolean,
+  map: ParameterScope.(KtParameter) -> Transform<KtParameter>
+) : ExtensionPhase =
+  quote(match, map) { ParameterScope(it) }
+
+/**
+ * A template destructuring [Scope] for a [KtParameter]
+ */
+class ParameterScope(
+  override val value: KtParameter,
+  val modality: Name? = value.modalityModifierType()?.value?.let(Name::identifier),
+  val visibility: Name? = value.visibilityModifierType()?.value?.let(Name::identifier),
+  val name: Name? = value.nameAsName,
+  val `@annotationEntries`: ScopedList<KtAnnotationEntry> = ScopedList(value.annotationEntries),
+  val type: Scope<KtTypeReference> = Scope(value.typeReference),
+  val `(typeParameters)`: ScopedList<KtTypeParameter> = ScopedList(prefix = "<", value = value.typeParameters, postfix = ">"),
+  val defaultValue: KtExpression? = value.defaultValue
```

Comment:
@bloderxd I think we're missing some items here.  The way to think about AST elements is by the elements themselves. For example, for a parameter AST element, we might be interested in `hasValOrVar` and also `isVarArg`.

@raulraja I'm a little unsure about `KtParameter::getValueParameters` and so on - I would say we should add of lot of these other options as well, but it looks like it's simply not implemented in the class since a lot of them return `null` or `emptyList()`

Reply:
@ahinchman1 functions like `hasValOrVar`, `isVarArg` are already provided by  KtParameter. It would add little as we have `KtParameter` in Scope at `match`.